File: algorithms/python/_003_LongestSubstringWithoutRepeatingCharacters/_003_LongestSubstringWithoutRepeatingCharacters_practise.py
class Solution:
    # start: 2019-07-22 18:34:48
    # end : 2019-07-22 18:46:24
    def lengthOfLongestSubstring(self, s):
        length, i, res = len(s), 0, 0
        map = {}
        for j in range(length):
            if s[j] in map:
                i = max(map.get(s[j]), i)

            res = max(res, j - i + 1)
            map[s[j]] = j + 1
        return res


if __name__ == "__main__":
    so = Solution()
    s = "abcabcbb"
    res = so.lengthOfLongestSubstring(s)
    print(res)

File: algorithms/cpp/_003_LongestSubstringWithoutRepeatingCharacters/_003_LongestSubstringWithoutRepeatingCharacters_practise.cpp
#include <iostream>
#include <vector>
#include <set>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string &s) {
        int ans = 0;
        int i = 0, j = 0;
        set<char> set;
        int length = s.length();
        while (i <= j && j < length) {
            if (set.find(s[j]) == set.end()) {
                set.insert(s[j++]);
                ans = std::max(ans, j - i);
            } else
                set.erase(s[i++]);
        }
        return ans;
    }
};

int main() {
//    2019-09-06 17:32:47
//2019-09-06 17:53:21
    auto *so = new Solution();
    string s = "pwwkew";
    int res = so->lengthOfLongestSubstring(s);
    cout << res << endl;
    delete so;
    return 0;
}
